packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts

```diff
@@ -0,0 +1,173 @@
+import { isRTL as isRTLFn } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import * as styles from '../styles';
+
+// Possible values:
+// - 'INITIAL': the initial state
+// - 'ANIMATING_IN': start enter animation
+// - 'IN': enter animation has ended
+// - 'ANIMATING_OUT': start exit animation
+// - 'OUT': the exit animation has ended
+type AnimationStatus =
+	| 'INITIAL'
+	| 'ANIMATING_IN'
+	| 'IN'
+	| 'ANIMATING_OUT'
+	| 'OUT';
+
+const isEnterAnimation = (
+	animationDirection: 'forwards' | 'backwards',
+	animationStatus: AnimationStatus,
+	animationName: string
+) =>
+	animationStatus === 'ANIMATING_IN' &&
+	animationName === styles.ANIMATION_END_NAMES[ animationDirection ].in;
+
+const isExitAnimation = (
+	animationDirection: 'forwards' | 'backwards',
+	animationStatus: AnimationStatus,
+	animationName: string
+) =>
+	animationStatus === 'ANIMATING_OUT' &&
+	animationName === styles.ANIMATION_END_NAMES[ animationDirection ].out;
+
+export function useScreenAnimatePresence( {
+	isMatch,
+	skipAnimation,
+	isBack,
+	onAnimationEnd,
+	screenEl,
+	setWrapperHeight,
+}: {
+	isMatch: boolean;
+	skipAnimation: boolean;
+	isBack?: boolean;
+	onAnimationEnd?: React.AnimationEventHandler< Element >;
+	screenEl?: HTMLElement | null;
+	setWrapperHeight?: ( height: number | undefined ) => void;
+} ) {
+	const isRTL = isRTLFn();
+	const animationTimeoutRef = useRef< number >();
+	const prefersReducedMotion = useReducedMotion();
+
+	const [ animationStatus, setAnimationStatus ] =
+		useState< AnimationStatus >( 'INITIAL' );
+
+	const wasMatch = usePrevious( isMatch );
+
+	const screenHeightRef = useRef< number | undefined >();
+
+	// Start enter and exit animations when the screen is selected or deselected.
+	// The animation status is set to `*_END` immediately if the animation should
+	// be skipped.
+	useLayoutEffect( () => {
+		if ( ! wasMatch && isMatch ) {
+			screenHeightRef.current = undefined;
+			setAnimationStatus(
+				skipAnimation || prefersReducedMotion ? 'IN' : 'ANIMATING_IN'
+			);
+		} else if ( wasMatch && ! isMatch ) {
+			screenHeightRef.current = screenEl?.offsetHeight;
+			setAnimationStatus(
+				skipAnimation || prefersReducedMotion ? 'OUT' : 'ANIMATING_OUT'
+			);
+		}
+	}, [ isMatch, wasMatch, skipAnimation, prefersReducedMotion, screenEl ] );
+
+	// When starting an animation, set the wrapper height to the screen height,
+	// to prevent layout shifts during the animation.
+	useEffect( () => {
+		if ( animationStatus === 'ANIMATING_OUT' ) {
+			setWrapperHeight?.( screenHeightRef.current ?? 0 );
+		} else if ( animationStatus === 'OUT' ) {
+			setWrapperHeight?.( undefined );
+		}
+	}, [ screenEl, animationStatus, setWrapperHeight ] );
+
+	// Fallback timeout to ensure the screen is removed from the DOM in case the
+	// `animationend` event is not triggered.
+	useEffect( () => {
+		if ( animationStatus === 'ANIMATING_OUT' ) {
+			animationTimeoutRef.current = window.setTimeout( () => {
```

Comment:
`animationTimeout` can be a local variable in the effect function, no need for an external ref.

Also, should this fallback timeout also change the state from `ANIMATING_IN` to `IN`? The `onScreenAnimationEnd` does that. If I understand correctly, the idea is that if the `animationEnd` callback doesn't run for some reason, the fallback callback will do its job anyway after some timeout.

Reply:
Yes, that's correct — I forgot to update the timeout in the last refactor

Reply:
https://github.com/WordPress/gutenberg/pull/64777/commits/0a697e2d89981c8645481cbcfd015c5c9b014952